Add all tasks for Module 4 with updated  requirements.txt

GeorgiiDermenzhy opened the pull request:
In this pull request I've added ...

<!-- Please mention all relevant module/task numbers. -->

# Pull Request Checklist

<!-- Please keep this section. It will make the maintainer's life easier. -->

* [x] I've checked code quality with flake8 and pylint tools
* [x] I've checked compliance with Python docstring conventions with pydocstyle tool
* [x] All third party packages are added to requirements.txt

GeorgiiDermenzhy committed:
Add all tasks for Module 4 with updated  requirements.txt

musk-ox commented on tasks/Module 4/task_6.py:
```diff
@@ -0,0 +1,26 @@
+    dict_file = askopenfilename(title='Select encrypted Dictionary file')
+    dict_file_path = Path(dict_file)
+
+    with open(dict_file_path, "r") as dict_file_content:
```
If you like you can also open multiple files in one `with` statement:
```python
with open(dict_file_path, "r") as dict_file_content, open(pwd_file_path.name, 'rb') as pdf_content:
```

GeorgiiDermenzhy replied:
The code is changed based on above

musk-ox commented on tasks/Module 4/task_6.py:
```diff
@@ -0,0 +1,26 @@
+        with open(pwd_file_path.name, 'rb') as pdf_content:
+            pdf_reader = PyPDF2.PdfFileReader(pdf_content)
+            for word in dict_file_content:
+                if pdf_reader.decrypt(word) or pdf_reader.decrypt(word.lower()):
```
With such condition your script does not know which word (lower vs upper) is the password. I am afraid that it requires two separate conditions.

GeorgiiDermenzhy replied:
The if statement has been splited into `if` for uppercase word and `elif` for lowercase

musk-ox commented on tasks/Module 4/task_1.py:
```diff
@@ -0,0 +1,41 @@
+                    if word in WORDS_TO_REPLACE:
+                        input_word = pyip.inputStr(f"Please provide {word}:\n")
+                        new_text.append(input_word)
+                    elif word[:len(word) - 1] in WORDS_TO_REPLACE:
```
I suppose that it is for case when word is ended with punctuation mark. If you want you can think about splitting input text using some [smart regular expression](https://stackoverflow.com/questions/367155/splitting-a-string-into-words-and-punctuation), so you won't need this condition.

GeorgiiDermenzhy replied:
@musk-ox right now my idea is to introduce new variable `slitted_sentence = re.findall(r"[\w']+|\.", line)` and then iterate over it, but as a result i'll have a string with a space before the dots(at least if i'll try to `output_file.write(" ".join(new_text))` )

if there a way we could simply connect the data? As right now i see it as we replacing one `elif` with a loop to concatenate it all back :)

musk-ox replied:
Good point @GeorgiiDermenzhy. To overcome this we can modify the regular expression by adding `|\s` to include also spaces:
```python
slitted_sentence = re.findall(r"[\w']+|\.|\s", line)
```
Of course then you will need to join by empty string:
```python
output_file.write("".join(new_text)))
```

GeorgiiDermenzhy replied:
@musk-ox works perfectly as always, thanks :)

musk-ox commented on tasks/Module 4/task_1.py:
```diff
@@ -0,0 +1,41 @@
+            input_file.write("The ADJECTIVE panda walked to the NOUN and then VERB. "
+                             "A nearby NOUN was unaffected by these events.")
+
+    with open(input_file_path, "r") as input_file:
```
You can combine this into one with statement:
```python
with open(input_file_path, "r") as input_file, open(output_file_path, "w") as output_file:
```

GeorgiiDermenzhy replied:
The code is changed based on above

musk-ox commented on tasks/Module 4/task_1.py:
```diff
@@ -0,0 +1,41 @@
+                        new_text.append(word)
+            output_file.write(" ".join(new_text))
+
+    with open(output_file_path, "r") as input_file:
```
There is no need to read from output file, because you already have this in accessible as `" ".join(new_text)`.

GeorgiiDermenzhy replied:
Opening of the files is removed and replaced with `print("".join(new_text))`

musk-ox commented on tasks/Module 4/task_2.py:
```diff
@@ -0,0 +1,39 @@
+from tkinter.filedialog import askdirectory
+
+
+def filling_in_the_gaps():
```
It does not work as expected for some cases. Please test it with following input files: `spam001.txt`, `spam002.txt`, `spam004.txt`, `spam007.txt`. Expected result is `spam001.txt`, `spam002.txt`, `spam003.txt`, `spam004.txt`, but instead I got `spam002.txt`, `spam003.txt`, `spam004.txt`.

GeorgiiDermenzhy replied:
@musk-ox strange bu it works for me. I checked several times and the fiels from `spam001.txt, spam002.txt, spam004.txt, spam007.txt` got renamed to `spam001.txt, spam002.txt, spam003.txt, spam004.txt`

wluzarowski-epam replied:
Weird, with old version of code I was getting wrong results. Maybe it was because of lack of sorting and `os.listdir` return results in different order depending on OS. Anyway, it works now 👍

musk-ox commented on tasks/Module 4/task_2.py:
```diff
@@ -0,0 +1,39 @@
+
+    os.chdir(dir_path)
+
+    for file in os.listdir(dir_path):
```
I just leave here advice from our program chat ;)

Please keep in mind that `os.listdir` or `os.scandir` return directory entries in arbitrary order, so make sure that your algorithm doesn't rely on that order. Otherwise you may get file names without gaps in numbering, but their content will be mixed  The simplest way to be safe is to sort input files by name: https://docs.python.org/3/howto/sorting.html

GeorgiiDermenzhy replied:
it was hard to miss that, but i somehow did xD
changed to `for file in sorted(os.listdir(dir_path))`

musk-ox replied:
@GeorgiiDermenzhy did you commit this change? ;)

GeorgiiDermenzhy replied:
@musk-ox, just finished all the local changes and pushed :)

musk-ox commented on tasks/Module 4/task_2.py:
```diff
@@ -0,0 +1,39 @@
+    for file in os.listdir(dir_path):
+        file_path = Path(file)
+        files_number = file_path.stem[4:7]
+        expected_numbers = str(hundred) + str(dozen) + str(digit)
```
It is an interesting algorithm, but Python can do it all for you :) All you need is a valid number of file and [str.zfill method](https://docs.python.org/3/library/stdtypes.html#str.zfill) or f-string `f'{file_number:03}'`.

GeorgiiDermenzhy replied:
Algoritm replaced with introducing new variable: `expected_number = str(iter).zfill(3)`

musk-ox commented on tasks/Module 4/task_7.py:
```diff
@@ -0,0 +1,33 @@
+                with open(destination, "w", newline='') as csv_file:
+                    csv_writer = csv.writer(csv_file)
+                    for row in ws.rows:
+                        row_data = []
```
When you have a list or other sequence and want to create new list by performing some operation on every element, you can use [list comprehension](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions):
```python
row_data = [cell.value for cell in row]
```
It is pretty convenient and you can also use conditions to filter out some data.

GeorgiiDermenzhy replied:
Thanks @musk-ox, ive changed this For loop with using lict comprehensions

GeorgiiDermenzhy committed:
Module 4 task 6: open files in one line and one "with"

GeorgiiDermenzhy committed:
Module 4 task 6: split if statement for pdf decryption into two conditions(lower and upper case word)

GeorgiiDermenzhy committed:
Module 4 task 1: Change of looping through line.split() to splitted_sentence that has list of of words splitted by using regular expression

GeorgiiDermenzhy committed:
Module 4 task 2: simplify the function with introducing of expected_number variable with expected file index number

GeorgiiDermenzhy committed:
Module 4 task 7: Change for loop to list comprehensions

wluzarowski-epam commented on .github/requirements.txt:
```diff
@@ -2,4 +2,13 @@ pylint
+pyinputplus
+pathlib
+askdirectory
+os
```
There is no need for putting built-in modules into this file. Its purpose is to list 3rd party libraries, so other people can install the same packages that you have. Please remove from here:

- pathlib
- os
- csv
- shutil

GeorgiiDermenzhy replied:
built-in modules removed

wluzarowski-epam commented on tasks/Module 4/task_2.py:
```diff
@@ -0,0 +1,32 @@
+        if files_number != expected_number:
+            os.rename(file, file_prefix + expected_number + ".txt")
+            iter += 1
+        else:
```
Two points:
- You are incrementing `iter` in both cases, so it can be pulled outside the `if...else` statement. In fact you only need `if` part.
- Python can handle it for you :) If you have a for loop and a variable that is incremented in each step, then use `enumerate` instead:
```python
for index, file in enumerate(sorted(os.listdir(dir_path)), start=1):
```
Please noticed that I renamed `iter` to `index` in the example, because `iter` is name of built-in function.

GeorgiiDermenzhy replied:
`else` removed

`enumerate` used in for loop

wluzarowski-epam commented on tasks/Module 4/task_6.py:
```diff
@@ -0,0 +1,29 @@
+                break
+            else:
+                pdf_reader.decrypt(word.lower())
+                print(word)
```
Shouldn't it be `print(word.lower())`?

GeorgiiDermenzhy replied:
Changed as proposed

GeorgiiDermenzhy committed:
Module 4 task_2, task_6:

task_2: Remove else statement
		Use enumerate in for loop

task_6: Change print(word) to print(word.lower())
